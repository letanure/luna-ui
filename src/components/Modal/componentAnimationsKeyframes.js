import { keyframes } from 'styled-components';

export const componentAnimationsKeyframes = {
  fadeIn: keyframes`
    from {
      opacity: 0;
      visibility: hidden;
    }

    to {
      opacity: 1;
      visibility: visible;
    }
  `,
  fadeOut: keyframes`
    from {
      opacity: 1;
      visibility: visible;
    }

    to {
      opacity: 0;
      offset: 0;
      visibility: hidden;
    }
  `,
  bounceIn: keyframes`
    0% {
      offset: 0;
      opacity: 0;
      transform: scale3d(.3, .3, .3);
    }

    20% {
      transform: scale3d(1.1, 1.1, 1.1);
    }

    40% {
      transform: scale3d(.9, .9, .9);
    }

    60% {
      opacity: 1;
      transform: scale3d(1.03, 1.03, 1.03);
    }

    80% {
      transform: scale3d(.97, .97, .97);
    }

    100% {
      opacity: 1;
      transform: scale3d(1, 1, 1);
    }
  `,
  bounceOut: keyframes`
    0% {
      opacity: 1;
      transform: 'none';
    }

    20% {
      opacity: 1;
      transform: scale3d(.9, .9, .9);
    }

    50% {
      opacity: 1;
      transform: scale3d(1.1, 1.1, 1.1);
    }

    55% {
      opacity: 1;
      transform: scale3d(1.1, 1.1, 1.1);
    }

    80% {
      transform: scale3d(.97, .97, .97);
    }

    100% {
      opacity: 0;
      transform: scale3d(.3, .3, .3);
    }
  `
};
