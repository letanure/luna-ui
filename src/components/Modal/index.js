import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import FontAwesomeIcon from '@fortawesome/react-fontawesome';

import hex2Rgba from '../../ultils/hex2Rgba';

import { componentAnimationsKeyframes } from './componentAnimationsKeyframes';

const firstLoad = true;

class ModalComponent extends React.Component {
  constructor(props) {
    super(props);

    this.state = { active: props.active };
    this.handleCloseModalAction = this._handleCloseModalAction.bind(this);
  }

  _handleCloseModalAction() {
    this.setState({ active: false });

    // eslint-disable-next-line no-undef
    this.desableTimeout = setTimeout((_) => {
      this.props.onSetModalAsDisabled();

      // eslint-disable-next-line no-undef
      clearTimeout(this.desableTimeout);
    }, 300);
  }

  _getFadeAnimation() {
    if (this.state.active && firstLoad) {
      return `animation: ${componentAnimationsKeyframes.fadeIn} .25s ease-in;`;
    }
    return `animation: ${componentAnimationsKeyframes.fadeOut} .25s ease-in;`;
  }

  _getBounceAnimation() {
    if (this.state.active) {
      return `animation: ${componentAnimationsKeyframes.bounceIn} .7s .25s cubic-bezier(0.215, 0.610, 0.355, 1.000);`;
    } else if (!this.state.active) {
      return `animation: ${componentAnimationsKeyframes.bounceOut} .7s;`;
    }
  }

  render() {
    const { className } = this.props;

    const ModalWrapper = styled.section`
      width: 100%;
      height: 100%;
      padding: 5%;
      top: 0;
      left: 0;
      z-index: 10000;
      position: fixed;
      overflow: scroll;
      background-color: ${hex2Rgba('#333333', 0.9)};

      // animation
      opacity: 0;
      visibility: hidden;
      ${this._getFadeAnimation()}
      animation-fill-mode: forwards;
    `;

    const ModalContent = styled.div`
      width: 100%;
      max-width: 600px;
      padding: 20px;
      margin: 0 auto;
      position: relative;
      background-color: #f0f0f0;
      top: 20%;
      transform: translateY(-20%);
      margin-bottom: 200px;

      // animation
      opacity: 0;
      ${this._getBounceAnimation()}
      animation-fill-mode: forwards;

      ${(props) =>
    `border-top: 2px solid ${props.theme.buttonColors.primary}`
};
    `;

    const CloseModal = styled.button`
      border: none;
      outline: none;
      cursor: pointer;
      width: 50px;
      height: 50px;
      font-size: 20px;
      color: #fff;
      position: absolute;
      top: -52px;
      right: 0;
      transition: opacity linear 0.25s;

      ${(props) =>
    `background-color: ${props.theme.buttonColors.primary}`
};

      &:hover {
        opacity: 0.8;
      }
    `;

    return (
      <ModalWrapper>
        <ModalContent>
          <CloseModal
            className={className}
            onClick={this._handleCloseModalAction.bind(this)}>
            <FontAwesomeIcon icon="times" />
          </CloseModal>
          {this.props.children}
        </ModalContent>
      </ModalWrapper>
    );
  }
}

ModalComponent.propTypes = {
  active: PropTypes.bool.isRequired,
  onSetModalAsDisabled: PropTypes.func.isRequired,
  children: PropTypes.node.isRequired,
  className: PropTypes.string
};

export default ModalComponent;
