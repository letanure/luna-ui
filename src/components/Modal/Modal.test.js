import React from 'react';
import sinon from 'sinon';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

import theme from '../../config/defaultTheme';
import ThemeProvider from '../Theme/index';

import Modal from './index.js';

describe('<Modal />', () => {
  let onSetModalAsDisabled; let activeModal; let deactiveModal;

  beforeEach(() => {
    onSetModalAsDisabled = sinon.spy();

    activeModal =
      <ThemeProvider theme={theme}>
        <Modal active={true} onSetModalAsDisabled={onSetModalAsDisabled}><h3>My Modal2</h3></Modal>
      </ThemeProvider>;

    deactiveModal =
      <ThemeProvider theme={theme}>
        <Modal active={false} onSetModalAsDisabled={onSetModalAsDisabled}><h3>My Modal 1</h3></Modal>
      </ThemeProvider>;
  });

  test('onSetModalAsDisabled', () => {
    jest.useFakeTimers();
    const wrapper = mount(activeModal);
    wrapper.find('button').at(0)
      .simulate('click');
    jest.runAllTimers();
    expect(window.setTimeout).toHaveBeenCalledTimes(1);
    expect(onSetModalAsDisabled).toHaveProperty('callCount', 1);
  });

  test('toMatchSnapshot: Active Modal', () => {
    const tree = renderer
      .create(activeModal)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('toMatchSnapshot: Deactive Modal', () => {
    const tree = renderer
      .create(deactiveModal)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
