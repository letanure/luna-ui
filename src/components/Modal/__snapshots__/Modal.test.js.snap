// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal /> toMatchSnapshot: Active Modal 1`] = `
<div>
  <section
    className="sc-bZQynM fTdDRP"
  >
    <div
      className="sc-gzVnrw ilmxXx"
    >
      <button
        className="sc-htoDjs kmfhFx"
        onClick={[Function]}
      />
      <h3>
        My Modal2
      </h3>
    </div>
  </section>
</div>
`;

exports[`<Modal /> toMatchSnapshot: Deactive Modal 1`] = `
<div>
  <section
    className="sc-dnqmqq jisnn"
  >
    <div
      className="sc-iwsKbI kOdJPe"
    >
      <button
        className="sc-gZMcBi cfYjKE"
        onClick={[Function]}
      />
      <h3>
        My Modal 1
      </h3>
    </div>
  </section>
</div>
`;
